Generate: use `GenerationConfig` as the basis for `.generate()` parametrization (#20388)
huggingface/transformers: docs/source/en/main_classes/text_generation.mdx

- TensorFlow [`~generation.TFGenerationMixin.generate`] is implemented in [`~generation.TFGenerationMixin`].
- Flax/JAX [`~generation.FlaxGenerationMixin.generate`] is implemented in [`~generation.FlaxGenerationMixin`].

> Regardless of your framework of choice, you can parametrize the generate method with a [`~generation.GenerationConfig`]

Reviewer: ```suggestion
Regardless of your framework of choice, you can parameterize the generate method with a [`~generation.GenerationConfig`]
```

class instance. Please refer to this class for the complete list of generation parameters, which control the behavior
of the generation method.

All models have a default generation configuration that will be used if you don't provide one. If you have a loaded
model instance `model`, you can inspect the default generation configuration with `model.generation_config`. If you'd
like to set a new default generation configuration, you can create a new [`~generation.GenerationConfig`] instance and
store it with `save_pretrained`, making sure to leave its `config_file_name` argument empty.

```python
from transformers import AutoModelForCausalLM, GenerationConfig

model = AutoModelForCausalLM.from_pretrained("my_account/my_model")

# Inspect the default generation configuration
print(model.generation_config)

# Set a new default generation configuration
generation_config = GenerationConfig(
    max_new_tokens=50, do_sample=True, top_k=50, eos_token_id=model.config.eos_token_id
)
generation_config.save_pretrained("my_account/my_model", push_to_hub=True)
```

<Tip>

If you inspect a serialized [`~generation.GenerationConfig`] file or print a class instance, you will notice that

> default values are omitted. Some attributes, like `max_new_tokens`, have a conservative default value, to avoid running

Reviewer: `max_length` as well no?
```suggestion
default values are omitted. Some attributes, like `max_new_tokens` and `max_length`, have a conservative
```

> default values are omitted. Some attributes, like `max_new_tokens`, have a conservative default value, to avoid running

Reviewer: `max_new_tokens` defaults to `None`. I think it's `max_length` that defaults to 20 no? 
Honestly I don't think we can never remove `max_length` from the config and also not remove the default. It's simply used too much and `max_new_tokens` is harder to serve as a good default.

So maybe just say "max_length" here?
```suggestion
default values are omitted. Some attributes, like `max_length`, have a conservative default value, to avoid running
```

(nit) Also `top_k` has a default set to 50 which could be mentioned here.

into resource limitations. Make sure you double-check the defaults in the documentation.

</Tip>

You can also store several generation parametrizations in a single directory, making use of the `config_file_name`
argument in `save_pretrained`. You can latter instantiate them with `from_pretrained`. This is useful if you want to
store several generation configurations for a single model (e.g. one for creative text generation with sampling, and
other for summarization with beam search).

```python
from transformers import GenerationConfig

# Create a new generation configuration for later use

> my_awesome_config = GenerationConfig(num_beams=8, do_sample=False, bad_words_ids=[[42, 43, 44]], eos_token_id=2)

Reviewer: Maybe use some "real-world" example? E.g. using T5 for both translation and QA?

my_awesome_config.save_pretrained(
    "my_account/my_model", config_file_name="awesome_generation_config.json", push_to_hub=True
)

# Restore the generation configuration
generation_config = GenerationConfig.from_pretrained(
    "my_account/my_model", config_file_name="awesome_generation_config.json"
)

> ```

Reviewer 2: Add here a line to show how to use it `generate` maybe?
